state-commitments/optimized/history_commitment.go

```diff
@@ -0,0 +1,489 @@
+	m := len(leaves)
+	depth := int(math.Log2(float64(limit)))
+	for j := 0; j < depth; j++ {
+		// Check to ensure we don't access out of bounds.
+		for i := 0; i < m/2 && 2*i+1 < m; i++ {
+			if _, err := h.keccak.Write(leaves[2*i][:]); err != nil {
+				return common.Hash{}, err
+			}
+			if _, err := h.keccak.Write(leaves[2*i+1][:]); err != nil {
+				return common.Hash{}, err
+			}
+			if _, err := h.keccak.Read(leaves[i][:]); err != nil {
+				return common.Hash{}, err
+			}
+			h.keccak.Reset()
+		}
+		if m&1 == 1 && m-1 < len(leaves) {
+			// Check to ensure m-1 is a valid index.
+			if _, err := h.keccak.Write(leaves[m-1][:]); err != nil {
+				return common.Hash{}, err
+			}
+			if j < len(fillers) { // Check to prevent index out of range for fillers.
+				if _, err := h.keccak.Write(fillers[j][:]); err != nil {
+					return common.Hash{}, err
+				}
+			} else {
+				// Handle the case where j is out of range for fillers.
+				return common.Hash{}, errors.New("insufficient fillers")
+			}
+			if _, err := h.keccak.Read(leaves[(m-1)/2][:]); err != nil {
+				return common.Hash{}, err
+			}
+			h.keccak.Reset()
+		}
+		m = (m + 1) / 2
+	}
+	return leaves[0], nil
+}
+
+// computeVirtualSparseTree returns the htr of a hashtree where the first layer
+// is passed as leaves, the completed with the last leaf until it reaches
+// virtual and finally completed with zero hashes until it reaches limit.
+// limit is assumed to be either 0 or a power of 2 which is greater or equal to
+// virtual. If limit is zero it behaves as if it were the smallest power of two
+// that is greater or equal than virtual.
+//
+// The algorithm is split in three different logic parts:
+//
+//  1. If the virtual length is less than half the limit (this can never happen
+//     in the first iteration of the algorithm), then the first half of the tree
+//     is computed by recursion and the second half is a zero hash of a given
+//     depth.
+//  2. If the leaves all fit in the first half, then we can optimize the first
+//     half to being a simple sparse tree, just that instead of filling with zero
+//     hashes we fill with the precomputed virtual hashes. This is the most common
+//     starting scenario. The second part is computed by recursion.
+//  3. If the leaves do not fit in the first half, then we can compute the first half of
+//     the tree as a normal full hashtree. The second part is computed by recursion.
+func (h *HistoryCommitter) computeVirtualSparseTree(leaves []common.Hash, virtual, limit uint64) (common.Hash, error) {
+	m := uint64(len(leaves))
+	if m == 0 {
+		return common.Hash{}, errors.New("nil leaves")
+	}
+	if virtual < m {
+		return common.Hash{}, fmt.Errorf("virtual %d should be >= num leaves %d", virtual, m)
+	}
+	var err error
+	if limit == 0 {
+		limit = nextPowerOf2(virtual)
+		// Check if m-1 is a valid index before accessing leaves[m-1]
+		if m > 0 {
+			n := 1
+			if virtual > m {
+				logValue := math.Log2(float64(virtual - m))
```

Comment:
Since the leaf fillers are being used also for proofs and this computation is fast (it's O(log(limit)) I think it's better to always do here the following
```suggestion
				logValue := math.Log2(float64(limit))
```
And remove all the logic involving `n`